### PR TITLE
vtol_att_control: use filtered airspeed for forward and backtransition acceptance & blending

### PR DESCRIPTION
**Pull request by oystub**
### Solved Problem
This solves an issue where spikes in airspeed values would cause oscillations in blending weights and in turn strong throttle oscillations, impairing transition performance.

### Solution
Airspeed measurements are filtered using a 1st order IIR low-pass filter which accounts for sample rate.
The filter time constant is controlled using parameter VT_ARSP_TAU.

In addition, it acts as a more robust replacement for the existing logic to prevent airspeed spikes affecting transition acceptance.

### CODE REVIEW

File: src/modules/logger/logged_topics.cpp
Comment (hermankolden) [inline src/modules/logger/logged_topics.cpp:57]: PX4 calls low pass filtered values "filtered", should use that convention

Comment (hermankolden) [reply src/modules/logger/logged_topics.cpp:57]: See e.g. https://docs.px4.io/main/en/msg_docs/BatteryStatus.html and https://docs.px4.io/main/en/msg_docs/TecsStatus.html

File: src/modules/vtol_att_control/vtol_type.h
Comment (hermankolden) [inline src/modules/vtol_att_control/vtol_type.h:229]: Is there a reason this float is a pointer? I don't see why

File: src/modules/vtol_att_control/vtol_att_control_main.h
Comment (hermankolden) [inline src/modules/vtol_att_control/vtol_att_control_main.h:165]: This seems like a strange location to define it

File: src/modules/vtol_att_control/vtol_att_control_main.cpp
Comment (hermankolden) [inline src/modules/vtol_att_control/vtol_att_control_main.cpp:470]:
```suggestion
			float alpha = dt_us / (_params.vt_aspd_tau * 1e6f);
```
No point actually calculating the full exponential (unless alpha might be large, then it's super important)

File: src/modules/vtol_att_control/standard.cpp
Comment (hermankolden) [inline src/modules/vtol_att_control/standard.cpp:153]: Have we done a risk analysis of this? What if something is wrong with the smoothing, so it back-transitions much too soon or much too late / not at all?

File: msg/airspeed_validated.msg
Comment (hermankolden) [inline msg/airspeed_validated.msg:14]: It seems weird to me to make an entirely separate topic where only 1 field is defined. Two options IMO:
1. Add a smoothed field to airspeed_validated
2. Define & smooth all fields in the new topic

File: src/modules/vtol_att_control/vtol_att_control_main.cpp
Comment (hermankolden) [inline src/modules/vtol_att_control/vtol_att_control_main.cpp:474]:
```suggestion
			_calibrated_airspeed_m_s_smoothed = _airspeed_validated.calibrated_airspeed_m_s;
```

File: src/modules/vtol_att_control/vtol_att_control_main.cpp
Comment (hermankolden) [inline src/modules/vtol_att_control/vtol_att_control_main.cpp:471]: Doesn't PX4 have existing functionality that you can use for low pass filtering? There are some gotchas (initialization, stale values, etc) which it would be nice to not re-invent

File: src/modules/vtol_att_control/standard.cpp
Comment (hermankolden) [inline src/modules/vtol_att_control/standard.cpp:194]: We should consider some safeguard in case the smoothed airspeed is weird. How do we know that the smoothed airspeed actually is healthy and represents what we want? I see the param limits are conservative, that might be enough

File: src/modules/vtol_att_control/standard.cpp
Comment (hermankolden) [inline src/modules/vtol_att_control/standard.cpp:191]: Same here, we should consider the risks of this value being wrong

File: src/modules/vtol_att_control/vtol_att_control_main.cpp
Comment (hermankolden) [inline src/modules/vtol_att_control/vtol_att_control_main.cpp:471]:
```suggestion
			_calibrated_airspeed_m_s_smoothed = _calibrated_airspeed_m_s_smoothed * (1-alpha) + _airspeed_validated.calibrated_airspeed_m_s * alpha;
```

File: src/modules/vtol_att_control/vtol_att_control_main.cpp
Comment (hermankolden) [reply src/modules/vtol_att_control/vtol_att_control_main.cpp:474]: A side effect of this if-else is that whenever the parameter value is below 0.1 and airspeed_validated gives publications, then _calibrated_airspeed_m_s_smoothed will perfectly track airspeed_validated, and be re-published as airspeed_smoothed, but there will be no smoothing. Is this intentional?

File: src/modules/vtol_att_control/standard.cpp
Comment (hermankolden) [inline src/modules/vtol_att_control/standard.cpp:274]: Whitespace

**Review by hermankolden — CHANGES_REQUESTED** [review]
The concept is sound, and I'm pretty sure this would be OK in a flight. However, since we alter the forward and back-transition behaviors where the drone is very sensitive to stall & crash, those two phases should be risk analyzed.

Regarding the implementation, my main concern is the introduction of a new topic rather than reuse of the airspeed_validated one with an extra field. This causes (probably) unnecessary overhead code and is harder to wrap my head around failure modes of.

Finally, I think we should copy existing low pass filters from PX4, e.g. lib/mathlib/math/filter/AlphaFilter.hpp which is used a bunch of places (I know it from the battery estimator with fields like battery_status.voltage_v_filtered and battery_status.current_a_filtered. This way we avoid unexpected errors with a custom implementation.

File: src/modules/vtol_att_control/vtol_att_control_main.cpp
Comment (hermankolden) [reply src/modules/vtol_att_control/vtol_att_control_main.cpp:471]: I recommend AlphaFilter from the maths library

File: src/modules/vtol_att_control/vtol_att_control_main.cpp
Comment (hermankolden) [inline src/modules/vtol_att_control/vtol_att_control_main.cpp:468]: Future Herman wonders where the exact value 0.1 comes from

File: src/modules/vtol_att_control/vtol_att_control_main.cpp
Comment (hermankolden) [reply src/modules/vtol_att_control/vtol_att_control_main.cpp:471]: Also: Future maintaining Herman wonders where the value 0.1 comes from and if it can be changed

File: src/modules/logger/logged_topics.cpp
Comment (oystub) [reply src/modules/logger/logged_topics.cpp:57]: Agree, should follow px4 convention, will update.

File: src/modules/vtol_att_control/vtol_type.h
Comment (oystub) [reply src/modules/vtol_att_control/vtol_type.h:229]: I don't like it either, but chosen to follow the coding pattern of the module. See variables above.

File: src/modules/vtol_att_control/vtol_att_control_main.h
Comment (oystub) [reply src/modules/vtol_att_control/vtol_att_control_main.h:165]: Can problably be moved, chose this to be right next to the topic which it tracks. Anyway, naming should be improved.

File: src/modules/vtol_att_control/standard.cpp
Comment (oystub) [reply src/modules/vtol_att_control/standard.cpp:153]: Time condition with `VT_B_TRANS_DUR` is a safeguard for late exit.

Since an LPFed signal will have a delay compared to raw measurements, early exit is less likely than with non-lpfed signal. In this case, top motors would immediately start, with a similar effect to quad-chute.

File: src/modules/vtol_att_control/vtol_att_control_main.cpp
Comment (oystub) [reply src/modules/vtol_att_control/vtol_att_control_main.cpp:470]: Agree!

File: msg/airspeed_validated.msg
Comment (oystub) [reply msg/airspeed_validated.msg:14]: This is a quick way of logging the value during initial test flights. Could probably have made it easier to review by adding an inline comment in addition to the commit message.

File: src/modules/vtol_att_control/vtol_att_control_main.cpp
Comment (oystub) [reply src/modules/vtol_att_control/vtol_att_control_main.cpp:474]: See above

File: src/modules/vtol_att_control/standard.cpp
Comment (oystub) [reply src/modules/vtol_att_control/standard.cpp:194]: What additional safeguards do you think should be introduced after LPFing the signal?
Assessment was that this is redundant when the input topic is already validated, and would just increase code size.
At least if using a standard lpf implementation, shouldn't this be safe?

Comment (oystub) [reply src/modules/vtol_att_control/standard.cpp:194]: For production implementation, the plan is to remove this entire code block, as an LPFed signal solved this issue by itself. Kept it in to remove change size. The slight risk of a transition timeout was considered acceptable for test flights.

File: src/modules/vtol_att_control/standard.cpp
Comment (oystub) [reply src/modules/vtol_att_control/standard.cpp:191]: See above

File: src/modules/vtol_att_control/vtol_att_control_main.cpp
Comment (oystub) [reply src/modules/vtol_att_control/vtol_att_control_main.cpp:471]: Good catch

Comment (oystub) [reply src/modules/vtol_att_control/vtol_att_control_main.cpp:471]: Yes, should choose a standard implementation, which allows values all the way down to zero safely.

File: src/modules/vtol_att_control/vtol_type.h
Comment (hermankolden) [reply src/modules/vtol_att_control/vtol_type.h:229]: I think the ones above are pointers because they are structs, and I even think they have to be pointers because of the uORB API. For a counterxample, see all the float vairables below (_ra_hrate, _ra_hrate_sp, etc)

File: msg/airspeed_validated.msg
Comment (hermankolden) [reply msg/airspeed_validated.msg:14]: It doesnt make sense to me that adding an additional topic is easier than modifying an existing one. How is it like that?

Comment (oystub) [reply msg/airspeed_validated.msg:14]: Because the existing topic is published from a different module.

File: src/modules/vtol_att_control/standard.cpp
Comment (hermankolden) [reply src/modules/vtol_att_control/standard.cpp:194]: Good point. Since this is an exponential filter, we know that the true value will always be reached in approx 3 x VT_ASPD_TAU = 15s, regardless of the starting point. There are two failure modes:

1. Delayed triggering of forward/backward transition acceptance because of 3xVT_ASPD_TAU delay.
1a) Delayed forward transition acceptance by up to 15sec: Safeguarded by VT_TRANS_TIMEOUT.
2b) Delayed back transition acceptance by up to 15sec: Safeguarded by VT_B_TRANS_DUR like you say.

2. Premature trigger of forward/backward transition due to lagging VT_ASPD_TAU. Does VTOL attitude controller run during FW? Otherwise, the LPF might be at e.g. 19 mps at the end of front transition, and thus start out low when back-transition starts.
2a) Premature acceptance of forward transition: Altitude drop, adaptive QuadChute.
2b) Premature acceptance of back transition: Spontenaous "QuadChute"

In conclusion I agree that these risks seem acceptable, based on what you wrote. Next time it's super useful if the failure analysis is written down

File: src/modules/vtol_att_control/standard.cpp
Comment (oystub) [inline src/modules/vtol_att_control/standard.cpp:266]: Don't know what I like better out of calling getter multiple times and making a temp variable.

File: src/modules/vtol_att_control/standard.cpp
Comment (oystub) [inline src/modules/vtol_att_control/standard.cpp:189]: The new filtered airspeed will replace this functionality for preventing spikes to trigger transition.

File: msg/airspeed_validated.msg
Comment (oystub) [reply msg/airspeed_validated.msg]: Replaced this with the vtol transition status.

File: src/modules/vtol_att_control/vtol_att_control_main.cpp
Comment (hermankolden) [inline src/modules/vtol_att_control/vtol_att_control_main.cpp:582]:
```suggestion
	if (dt <= 0) {
```

File: src/modules/vtol_att_control/vtol_att_control_main.cpp
Comment (hermankolden) [inline src/modules/vtol_att_control/vtol_att_control_main.cpp:596]:
```suggestion
	if (!sample.airspeed_sensor_measurement_valid || !PX4_ISFINITE(sample.calibrated_airspeed_m_s)) {
			_airspeed_filtered.update(sample.calibrated_airspeed_m_s);
	}
}
```

File: src/modules/vtol_att_control/standard.cpp
Comment (hermankolden) [reply src/modules/vtol_att_control/standard.cpp:266]: I'd say temp variable is more readable and robust, it's easier to see that they're the same variable. Maybe in theory the get_filtered_airspeed() could change between calls?

File: src/modules/vtol_att_control/vtol_att_control_main.cpp
Comment (hermankolden) [reply src/modules/vtol_att_control/vtol_att_control_main.cpp:596]: To understand why the extra complexity it worthy it: Why won't my suggested simplification above work?

**Review by hermankolden — CHANGES_REQUESTED** [review]
Looks better! I see you dropped the airspeed logging all together and replaced with the smoothing weights, which makes sense. I have a question about the loging wrapping the AlphaFilter, but you probably have a reason for it.

One inequality sign should be flipped.

Couple of things after testing:
* Reduce logging rate
* Consider temp variable instead of get_filtered_airspeed() for cleanliness

File: src/modules/vtol_att_control/standard.cpp
Comment (hermankolden) [reply src/modules/vtol_att_control/standard.cpp:189]: This should go into the commit description (ideally title too). Same with back-transition. When I read the commit, it would be awesome to see that FT acceptence, BT acceptance and FT blending are all affected.

File: src/modules/vtol_att_control/vtol_att_control_main.cpp
Comment (oystub) [reply src/modules/vtol_att_control/vtol_att_control_main.cpp:596]: We can skip the resets, since a long dt basically equals a reset.
However, we don't know in advance the airspeed sensor's sample rate, so must update the filter parameters.
Could simplify to this

```
	if (!sample.airspeed_sensor_measurement_valid || !PX4_ISFINITE(sample.calibrated_airspeed_m_s)) {
		if (!_airspeed_filtered_valid) {
			PX4_WARN("Received invalid calibrated airspeed");
		}
		_airspeed_filtered_valid = false;
		return;
	}

	if (sample.timestamp <= _airspeed_filtered_last_timestamp) {
		if (_airspeed_filtered_valid) {
			PX4_WARN("Airspeed timestamp not monotonic");
		}

		_airspeed_filtered_valid = false;
		return;
	}

	// We must update the filter parameters, since we don't know the sensor sample rate
	float dt = static_cast<float>(sample.timestamp - _airspeed_filtered_last_timestamp);
	_airspeed_filtered.setParameters(dt, _params.vt_aspd_tau);
	_airspeed_filtered.update(sample.calibrated_airspeed_m_s);
	_airspeed_filtered_last_timestamp = sample.timestamp;
	_airspeed_filtered_valid = true;
```

Comment (oystub) [reply src/modules/vtol_att_control/vtol_att_control_main.cpp:596]: We also need to output invalid data out if our input is invalid, since we want to trigger failsafes in the transition code if we loose airspeed.

File: src/modules/vtol_att_control/standard.cpp
Comment (oystub) [reply src/modules/vtol_att_control/standard.cpp:266]: The module is in a single thread, but not super-easy to see without a temp variable. Agree

**Review by hermankolden — APPROVED** [review]
Amazin'! Looks very good to me.

Remember to (maybe) reset logging rates before production

**Comment by oystub** [comment]
- [ ] Parameter should be renamed to VT_ARSP_TAU to fit in with VT_ARSP_BLEND and VT_ARSP_TRANS.

**Comment by oystub** [comment]
Squased and rebased onto main.

Analysis is done here: https://aviant.atlassian.net/wiki/spaces/TECHNICAL/pages/959381516/Analysis+Transition+with+airspeed+filtering

Need a final approval before merge.
https://github.com/aviant-tech/companion_computer/pull/167 Should be merged after this one to enable parameter monitoring for the VT_ARSP_TAU parameter.

**Pull request merged**
